### PR TITLE
#84 / SDCISA-3814 Queue pace (speed) evaluation feature added

### PR DESCRIPTION
**Pull request by Kusig**
Additional feature allowing to evaluate the message pace for a certain queue or group of queues defined by a regex filter. This allows to easily evaluate the pace as well for queues which are dynamically labelled by given header discrimators.

### CODE REVIEW

File: src/main/java/org/swisspush/redisques/handler/RedisquesHttpRequestHandler.java [hunk 1]
```diff
@@ -4,7 +4,43 @@
 import static org.swisspush.redisques.util.HttpServerRequestUtil.encodePayload;
 import static org.swisspush.redisques.util.HttpServerRequestUtil.evaluateUrlParameterToBeEmptyOrTrue;
 import static org.swisspush.redisques.util.HttpServerRequestUtil.extractNonEmptyJsonArrayFromBody;
-import static org.swisspush.redisques.util.RedisquesAPI.*;
```
Comment (linovifian) [inline src/main/java/org/swisspush/redisques/handler/RedisquesHttpRequestHandler.java:7]: ¯\_(ツ)_/¯

File: src/main/java/org/swisspush/redisques/util/QueueStatisticsCollector.java [hunk 2]
```diff
@@ -45,28 +49,79 @@
...
+     *
+     * NOTE: The pace values are held in memory only due to performance reasons. We don't want to
+     * store regularly a lot of values (all queues) for pace statistics only.
+     * This would no more work properly of course if there are multiple Redisques instances deployed.
+     * In such case, the client interested in such values must query each instance for itself.
+     *
+     * By default, the pace interval is configured to 1 Minute currently.
+     */
+    private void paceStatisticsScheduler(int paceIntervalTime){
+        if (paceIntervalTime<=0) {
```
Comment (dominik-cnx) [inline src/main/java/org/swisspush/redisques/util/QueueStatisticsCollector.java:80]: code formatting issue

File: src/main/java/org/swisspush/redisques/util/QueueStatisticsCollector.java [hunk 3]
```diff
@@ -45,28 +49,79 @@
...
+     * within the past time interval.
+     *
+     * NOTE: The pace values are held in memory only due to performance reasons. We don't want to
+     * store regularly a lot of values (all queues) for pace statistics only.
+     * This would no more work properly of course if there are multiple Redisques instances deployed.
+     * In such case, the client interested in such values must query each instance for itself.
+     *
+     * By default, the pace interval is configured to 1 Minute currently.
+     */
+    private void paceStatisticsScheduler(int paceIntervalTime){
```
Comment (dominik-cnx) [inline src/main/java/org/swisspush/redisques/util/QueueStatisticsCollector.java:79]: maybe name the argument paceIntervalTimeMinutes

File: src/main/java/org/swisspush/redisques/util/QueueStatisticsCollector.java [hunk 4]
```diff
@@ -45,28 +49,79 @@
...
         this.redisAPI = redisAPI;
         this.luaScriptManager = luaScriptManager;
         this.queuePrefix = queuePrefix;
+        this.vertx = vertx;
+        paceStatisticsScheduler(paceIntervalTime);
     }
 
     /**
-     * Does reset all statistics values of the given queue. In memory but as well the persisted
+     * This scheduled task collects and calculates the message pace per queue in the local system
```
Comment (dominik-cnx) [inline src/main/java/org/swisspush/redisques/util/QueueStatisticsCollector.java:69]: What is exactly the "message pace", there is also the term "queues pace". I think it is the same. Should we stick to one term and further describe / define it?

File: src/main/java/org/swisspush/redisques/util/QueueStatisticsCollector.java (hunk 3 above)
Comment (mcweba) [reply src/main/java/org/swisspush/redisques/util/QueueStatisticsCollector.java:79]: the unit for the `setPeriodic` method is milliseconds, not minutes. Either change the name of the parameter to `paceIntervalMs` or calculate to milliseconds before use.

File: src/main/java/org/swisspush/redisques/handler/RedisquesHttpRequestHandler.java (hunk 1 above)
Comment (Kusig) [reply src/main/java/org/swisspush/redisques/handler/RedisquesHttpRequestHandler.java:7]: As of our internal coding guidelines we don't want such wildcard imports. However, it was changed anyhow by the formatter and I let it do so for this reason.

File: src/main/java/org/swisspush/redisques/util/QueueStatisticsCollector.java (hunk 3 above)
Comment (Kusig) [reply src/main/java/org/swisspush/redisques/util/QueueStatisticsCollector.java:79]: Will rename all over with Ms to make it clear everywhere.

File: src/main/java/org/swisspush/redisques/util/QueueStatisticsCollector.java (hunk 4 above)
Comment (dominik-cnx) [reply src/main/java/org/swisspush/redisques/util/QueueStatisticsCollector.java:69]: OK, I think it makes sense, hope that it doesn't take too much time to adapt all the code.

**Review by mcweba — APPROVED** [review]

**Review by dominik-cnx — APPROVED** [review]

File: src/test/java/org/swisspush/redisques/handler/RedisquesHttpRequestHandlerTest.java [hunk 5]
```diff
@@ -190,6 +195,22 @@ protected void eventBusSend(JsonObject operation, Handler<AsyncResult<Message<Js
...
+    /**
+     * Helper method to register an eventbus queue message handler which consumes just all
+     * queue messages properly.
+     *
+     * Note: don't forget to unregister the given MessageConsumer after test completion. Else this
+     * might influence other tests later.
+     *
+     * @return The MessageConsumer used for later unregistration of the consumer.
+     */
+    private MessageConsumer registerQueueEventOkConsumer(){
```
Comment (ljucam) [inline src/test/java/org/swisspush/redisques/handler/RedisquesHttpRequestHandlerTest.java:207]: Couldn't you have used `@Before` and `@After`?

**Pull request merged**